vectorstores/pgvector/pgvector.go

```diff
@@ -0,0 +1,329 @@
+	ErrInvalidFilters             = errors.New("invalid filters")
+	ErrUnsupportedOptions         = errors.New("unsupported options")
+)
+
+// Store is a wrapper around the pgvector client.
+type Store struct {
+	embedder              embeddings.Embedder
+	conn                  *pgx.Conn
+	postgresConnectionURL string
+	embeddingTableName    string
+	collectionTableName   string
+	collectionName        string
+	collectionUUID        string
+	collectionMetadata    map[string]any
+	preDeleteCollection   bool
+}
+
+var _ vectorstores.VectorStore = Store{}
+
+// New creates a new Store with options.
+func New(ctx context.Context, opts ...Option) (Store, error) {
+	store, err := applyClientOptions(opts...)
+	if err != nil {
+		return Store{}, err
+	}
+	store.conn, err = pgx.Connect(ctx, store.postgresConnectionURL)
+	if err != nil {
+		return Store{}, err
+	}
+
+	if err = store.conn.Ping(ctx); err != nil {
+		return Store{}, err
+	}
+
+	if err = store.createVectorExtensionIfNotExists(ctx); err != nil {
+		return Store{}, err
+	}
+	if err = store.createCollectionTableIfNotExists(ctx); err != nil {
+		return Store{}, err
+	}
+	if err = store.createEmbeddingTableIfNotExists(ctx); err != nil {
+		return Store{}, err
+	}
+	if store.preDeleteCollection {
+		if err = store.RemoveCollection(ctx); err != nil {
+			return Store{}, err
+		}
+	}
+	if store.collectionUUID, err = store.createOrGetCollection(ctx); err != nil {
+		return Store{}, err
+	}
+	return store, nil
+}
+
+func (s Store) createVectorExtensionIfNotExists(ctx context.Context) error {
+	tx, err := s.conn.Begin(ctx)
+	if err != nil {
+		return err
+	}
+	// inspired by
+	// https://github.com/langchain-ai/langchain/blob/v0.0.340/libs/langchain/langchain/vectorstores/pgvector.py#L167
+	// The advisor lock fixes issue arising from concurrent
+	// creation of the vector extension.
+	// https://github.com/langchain-ai/langchain/issues/12933
+	// For more information see:
+	// https://www.postgresql.org/docs/16/explicit-locking.html#ADVISORY-LOCKS
+	if _, err := tx.Exec(ctx, "SELECT pg_advisory_xact_lock(1573678846307946495)"); err != nil {
+		return err
+	}
+	if _, err := tx.Exec(ctx, "CREATE EXTENSION IF NOT EXISTS vector"); err != nil {
+		return err
+	}
+	return tx.Commit(ctx)
+}
+
+func (s Store) createCollectionTableIfNotExists(ctx context.Context) error {
+	tx, err := s.conn.Begin(ctx)
+	if err != nil {
+		return err
+	}
+	// inspired by
+	// https://github.com/langchain-ai/langchain/blob/v0.0.340/libs/langchain/langchain/vectorstores/pgvector.py#L167
+	// The advisor lock fixes issue arising from concurrent
+	// creation of the vector extension.
+	// https://github.com/langchain-ai/langchain/issues/12933
+	// For more information see:
+	// https://www.postgresql.org/docs/16/explicit-locking.html#ADVISORY-LOCKS
+	if _, err = tx.Exec(ctx, "SELECT pg_advisory_xact_lock(1573678846307946494)"); err != nil {
+		return err
+	}
+	sql := fmt.Sprintf(`CREATE TABLE IF NOT EXISTS %s (
```

Comment:
rather than string-assemble this can you use arguments to Exec (throughout) -- this protects against sql injection attacks.

Reply:
I think Travis maybe  means that you shouldn't `Sprintf` into `package sql` calls like `Exec`, but rather use the `args` of each method, e.g. look at the example at https://pkg.go.dev/database/sql#example-Conn.ExecContext

`
result, err := conn.ExecContext(ctx, `UPDATE balances SET balance = balance + 10 WHERE user_id = ?;`, id)
`

See how `id` is passed into the SQL query?

Reply:
Yes that's what I meant.